homeassistant/core.py

```diff
@@ -762,11 +764,19 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
                 "Timed out waiting for shutdown stage 3 to complete, the shutdown will"
                 " continue"
             )
+            self._async_log_running_tasks(3)
+
         self.state = CoreState.stopped
 
         if self._stopped is not None:
             self._stopped.set()
 
+    def _async_log_running_tasks(self, stage: int) -> None:
+        """Log all running tasks."""
+        for task in self._tasks:
+            if not task.done():
```

Comment:
This shouldn't be possible, as there is a done callback to remove it from the set.

Reply:
Hm, I guess this is called with a `call_soon`: https://github.com/python/cpython/blob/main/Lib/asyncio/futures.py#L173

Reply:
> Hm, I guess this is called with a `call_soon`: [python/cpython@`main`/Lib/asyncio/futures.py#L173](https://github.com/python/cpython/blob/main/Lib/asyncio/futures.py?rgh-link-date=2023-02-25T04%3A15%3A37Z#L173)

I was worried about a race as I've seen them be done but the callback not yet fired.

Reply:
So `call_soon` callbacks are called [at the end of each iteration of the event loop](https://github.com/python/cpython/blob/main/Lib/asyncio/base_events.py#LL1928C18-L1933C47). Since the `async_block_till_done` waits for another async task, it should have executed the callback.

Reply:
81e845ba739475406fa7895a7fda60a9cad467c9